### ISSUE TITLE
Make build process free of warnings and errors

### ISSUE THREAD

**Issue by DougInAMug** [ForkAwesome/Fork-Awesome#311]
_The following comes from following the [build instructions](https://github.com/ForkAwesome/Fork-Awesome#building-fork-awesome) on Ubuntu 20.10._

## Building the fonts
`make -C src/icons` succeeds, with _"Using the last argument as keyword parameters is deprecated"_ warnings. These seem to arise from using [outdated ruby gems](https://bloggie.io/@kinopyo/how-to-fix-ruby-2-7-warning-using-the-last-argument-as-keyword-parameters-is-deprecated). Terminal dump:

```
$ make -C src/icons
make: Entering directory '/home/doug/ext/Fork-Awesome/src/icons'
Cleaning up SVG files...
npx svgo --quiet --pretty --config=svgo_config.json --folder=svg
Compiling Icons into a ForkAwesome fonts...
bundle exec fontcustom compile
/var/lib/gems/2.7.0/gems/json-1.8.6/lib/json/common.rb:155: warning: Using the last argument as keyword parameters is deprecated
/var/lib/gems/2.7.0/gems/json-1.8.6/lib/json/common.rb:155: warning: Using the last argument as keyword parameters is deprecated
      create  forkawesome
/var/lib/gems/2.7.0/gems/json-1.8.6/lib/json/common.rb:155: warning: Using the last argument as keyword parameters is deprecated
      create  forkawesome/forkawesome.ttf
              forkawesome/forkawesome.svg
              forkawesome/forkawesome.woff
              forkawesome/forkawesome.eot
              forkawesome/forkawesome.woff2
/var/lib/gems/2.7.0/gems/json-1.8.6/lib/json/common.rb:155: warning: Using the last argument as keyword parameters is deprecated
      create  forkawesome/forkawesome.css
              forkawesome/forkawesome-preview.html
/var/lib/gems/2.7.0/gems/json-1.8.6/lib/json/common.rb:155: warning: Using the last argument as keyword parameters is deprecated
Copying builds to root folder (package release)...
cp ./forkawesome/forkawesome.eot ../../fonts/forkawesome-webfont.eot
cp ./forkawesome/forkawesome.svg ../../fonts/forkawesome-webfont.svg
cp ./forkawesome/forkawesome.ttf ../../fonts/forkawesome-webfont.ttf
cp ./forkawesome/forkawesome.woff ../../fonts/forkawesome-webfont.woff
cp ./forkawesome/forkawesome.woff2 ../../fonts/forkawesome-webfont.woff2
done.
make: Leaving directory '/home/doug/ext/Fork-Awesome/src/icons'
```

## Building the website
`npm run build`, which aliases to `bundle exec jekyll build` within `package.json`, **fails** with:

```
lessc --clean-css="--compatibility=ie8" --source-map assets/fork-awesome/css/fork-awesome.css assets/fork-awesome/css/fork-awesome.min.css
undefinedError: Options have been provided but the plugin index.js does not support any options.
```

More specifically, the website html files are built, but there is no styling. Additionally, there is tsunami of the warnings mentioned above. Terminal dump:

```
...
...
...
/var/lib/gems/2.7.0/gems/jekyll-3.7.4/lib/jekyll/tags/include.rb:191: warning: Using the last argument as keyword parameters is deprecated
/var/lib/gems/2.7.0/gems/jekyll-3.7.4/lib/jekyll/tags/include.rb:191: warning: Using the last argument as keyword parameters is deprecated
Cleaning build dirs
rm -rf ./less/
rm -rf ./scss/
Compiling Less files
lessc assets/fork-awesome/less/fork-awesome.less assets/fork-awesome/css/fork-awesome.css
lessc --clean-css="--compatibility=ie8" --source-map assets/fork-awesome/css/fork-awesome.css assets/fork-awesome/css/fork-awesome.min.css
undefinedError: Options have been provided but the plugin index.js does not support any options.


make: *** [Makefile:30: build] Error 1
                    done in 7.875 seconds.
 Auto-regeneration: disabled. Use --watch to enable.
```

**Comment by DougInAMug** [comment]
I decided to create this dedicated issue, since it is not clearly expressed elsewhere. It seems partly related to #248 and possibly #267

**Comment by DougInAMug** [comment]
Investigative work: adding `lessc --version` to `src/doc/Makefile` before the point where the site-building error occurs outputs:
- `lessc 2.7.3 (Less Compiler) [JavaScript]` on Ubuntu 18.04 (VM) → succeeds
- `lessc 3.11.3 (Less Compiler) [JavaScript]` on Ubuntu 20.10 → fails

**Comment by DougInAMug** [comment]
Interestingly, `"less": "^3.11.3"` is listed in `devDependencies` in `package.json`...

**Comment by DougInAMug** [comment]
Ooh! Changing the `package.json` to `"less": "2.7.3"` now gives me a different error

```
lessc --version
lessc 2.7.3 (Less Compiler) [JavaScript]
lessc assets/fork-awesome/less/fork-awesome.less assets/fork-awesome/css/fork-awesome.css
lessc --clean-css="--compatibility=ie8" --source-map assets/fork-awesome/css/fork-awesome.css assets/fork-awesome/css/fork-awesome.min.css
Unable to interpret argument clean-css - if it is a plugin (less-plugin-clean-css), make sure that it is installed under or at the same level as less
```

**Comment by DougInAMug** [comment]
Specifying `less-plugin-clean-css` in `package.json` builds site successfully with `less` ar 2.7.3 and ^3.11.3! :tada:

**Comment by shinenelson** [comment]
This issue is a merge of two different issues that seems to be related but could have been tackled separately. But I believe most of the investigative work was to fix the warnings when building the website.

> It seems partly related to #248 

No, that one is a packaging issue. It is not a 'build' issue.

> and possibly #267

I was not able to reproduce #267 locally and was hoping to get back to it once we had streamlined the build system and build process.

**Comment by DougInAMug** [comment]
Yes, it could have been split, but we can think of it as an epic :wink: 

Agree to your comments on those issues.

**Comment by shinenelson** [comment]
> I ~am~ manually test[~ing~,ed] the build system on the following versions of ruby : 
> ```
>    ruby-2.4.9 [ x86_64 ]
>    ruby-2.5.1 [ x86_64 ]
>    ruby-2.5.8 [ x86_64 ]
>    ruby-2.6.0 [ x86_64 ]
>    ruby-2.6.1 [ x86_64 ]
>    ruby-2.6.2 [ x86_64 ]
>    ruby-2.6.3 [ x86_64 ]
>    ruby-2.6.4 [ x86_64 ]
>    ruby-2.6.5 [ x86_64 ]
>    ruby-2.6.6 [ x86_64 ]
>    ruby-2.7.0 [ x86_64 ]
>    ruby-2.7.1 [ x86_64 ]
>    ruby-2.7.2 [ x86_64 ]
> ```
>
> Ruby versions < 2.6 are already end-of-life ( EoL ). But I ~am~ still test[~ing~,ed] them as they still might be in use in different environments. Ruby 2.5 went EoL only 4 months ago ( 2021-03-31 ). And also Debian Buster has ruby 2.5.1.

( quoted from https://github.com/ForkAwesome/Fork-Awesome/pull/309/files#r677474300 )

The ruby warnings come from ruby-2.7.0 ( per the [blog post](https://bloggie.io/@kinopyo/how-to-fix-ruby-2-7-warning-using-the-last-argument-as-keyword-parameters-is-deprecated) linked from the issue description ). I manually validated it against the above versions too.

The easiest workaround to suppress the ruby warnings is to prepend `RUBYOPT='-W0'` to the `bundle exec jekyll` commands in `package.json` per [this Stack Overflow answer](https://stackoverflow.com/a/59594760); and my own validation in manual testing.

PS : the warnings are being emitted by the jekyll gem which is at version 3.7.4 in the repository. The 'tsunami' of warnings is emitted when attempting to [include](https://jekyllrb.com/docs/includes/) one file within another ( [offending line in source](https://github.com/jekyll/jekyll/blob/v3.7.4/lib/jekyll/tags/include.rb#L191) ) which is too common in jekyll. The issue were fixed with https://github.com/jekyll/jekyll/pull/7948 and released in https://github.com/jekyll/jekyll/releases/tag/v3.8.7. Also, [GitHub Pages currently uses version 3.9.0](https://pages.github.com/versions).

**Comment by shinenelson** [comment]
The `lessc` issues are kind of more convoluted.

TL;DR : the following patch should work : 
```diff
diff --git a/package.json b/package.json
index 933df7318..9720e7afc 100644
--- a/package.json
+++ b/package.json
@@ -23,9 +23,9 @@
   "license": "(OFL-1.1 AND MIT)",
   "dependencies": {},
   "devDependencies": {
-    "@wikimedia/less-plugin-clean-css": "^1.5.2",
     "all-contributors-cli": "^6.16.0",
     "less": "^3.11.3",
+    "less-plugin-clean-css": "less/less-plugin-clean-css#master",
     "svgo": "^1.3.2"
   },
   "engines": {

```

---

My initial objective was to isolate the problem. I removed `--clean-css="--compatibility=ie8"` from `src/doc/Makefile` and everything worked well. That confirmed that the problem was isolated to the `less-plugin-clean-css`.

```
lessc --clean-css="--compatibility=ie8" --source-map assets/fork-awesome/css/fork-awesome.css assets/fork-awesome/css/fork-awesome.min.css
undefinedError: Options have been provided but the plugin index.js does not support any options.
```
suggested that the `--clean-css` flag does not accept any options.

Next, I removed `="--compatibility=ie8"`. Everything worked well, again. Now, I had to figure why the `--compatibility=ie8` option did not work.

> Investigating deeper, I found that the 2 packages lock different major versions of `clean-css` which could possibly be the reason for different kind of behavior even though one is a fork of the other.
>
> ```
> +-- @wikimedia/less-plugin-clean-css@1.5.2
> | `-- clean-css@4.2.3
> |   `-- source-map@0.6.1 deduped
> ```
> ```
> +-- less-plugin-clean-css@1.5.1
> | `-- clean-css@3.4.28
> |   +-- commander@2.8.1
> |   | `-- graceful-readlink@1.0.1
> |   `-- source-map@0.4.4
> |     `-- amdefine@1.0.1
> ```

( quoted from https://github.com/ForkAwesome/Fork-Awesome/pull/312#issuecomment-887513443 )

There were [breaking changes](https://github.com/clean-css/clean-css#important-40-breaking-changes) between versions 3.x and 4.x of https://github.com/clean-css/clean-css.

Apparently, the version locked in `less-plugin-clean-css` had a [security vulnerability](https://www.npmjs.com/advisories?search=clean-css&version=3.4.28). `clean-css` was bumped to version 4.2.1 in `less-plugin-clean-css` via https://github.com/less/less-plugin-clean-css/pull/28 on October 13, 2019. However, there have not been any releases since then. And when they could not wait any longer for a [NPM release to resolve security vulnerability](https://github.com/less/less-plugin-clean-css/issues/29), Wikimedia Foundation [published](https://www.npmjs.com/package/@wikimedia/less-plugin-clean-css) [their](https://github.com/less/less-plugin-clean-css/issues/29#issuecomment-571128385) [fork](https://github.com/wikimedia/less-plugin-clean-css). That's probably how we inherited https://github.com/wikimedia/less-plugin-clean-css in https://github.com/ForkAwesome/Fork-Awesome/commit/8c73f16ee2d1f61f30e4651e6c67b013b2d74c90

However, even with no [significant modifications](https://github.com/less/less-plugin-clean-css/compare/master...wikimedia:master), Wikimedia's fork does not work with our build script where the `--clean-css="--compatibility=ie8"` is passed to `lessc`. I tested the same setup, but by hard-coding and using https://github.com/less/less-plugin-clean-css/tree/master and `npm run build` completed successfully. I cannot quite pin-point and explain what is causing the difference in behavior because both packages essentially have the same source code.

PS : version 4.2.0 of `clean-css` was the first version with the [security fix](https://github.com/clean-css/clean-css/blob/HEAD/History.md#420--2018-08-02) which was only [back-ported to version 4.1.11](https://github.com/clean-css/clean-css/blob/HEAD/History.md#4111--2018-03-06)

( aside : https://github.com/less is in a limbo-state with no maintainer; https://github.com/less/less.js/issues/3148 ).

**Comment by DougInAMug** [comment]
Wow! Great investigative work :clap: :detective: 

I can confirm your `less-plugin-clean-css` patch also works. I'm not familiar with the syntax you use, `less/less-plugin-clean-css#master` - is that tracking the master branch? My understanding is that the `^x.y.z` format chooses the most current, non-breaking release: isn't that safer? I trust you on this one, and am happy to merge either patch!

As for the ruby warnings, I can live with them either showing up, or suppressing them. What do you see as the most painless way to update the guilty components/are you 'on' this?

**Comment by shinenelson** [comment]
> I'm not familiar with the syntax you use, less/less-plugin-clean-css#master - is that tracking the master branch?

Yes, the notation is to use to the GitHub repository specified by `less/less-plugin-clean-css` and track the `master` branch. There are more ways to use [Git URLs as Dependencies](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#git-urls-as-dependencies), but I was suggesting the [GitHub URLs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#github-urls) method that has been supported since `npm` version 1.1.65.

> My understanding is that the ^x.y.z format chooses the most current, non-breaking release: isn't that safer?

It is the best ( and recommended ) approach when depending on _actively maintained_ dependencies. In this case, the [`less-plugin-clean-css` package](https://npmjs.com/package/less-plugin-clean-css) has been not been updated since 2015 and there has been a fork which has caused confusion for us.

Hence, we track the `master` branch of the upstream, canonical repository which has the same code as the [forked package](https://npmjs.com/package/@wikimedia/less-plugin-clean-css).

> As for the ruby warnings, I can live with them either showing up, or suppressing them. What do you see as the most painless way to update the guilty components/are you 'on' this?

For now, until we can bring the website to a decent state, let us suppress the warnings. I believe the warnings will automatically go away once we update the gems. And that will happen along with the website refactor.
